### PR TITLE
[#31] [#64] [Integrate] As a logged-in user, I can see the survey list on the Home screen

### LINKED ISSUE #31: [Integrate] As a logged-in user, I can see the survey list on the Home screen
## Why
After login, the user will see the survey list on the Home screen.

## Acceptance Criteria

The app will make an API request to get the survey list.
While the API is being called, the app will show the previously-cached data, if any. 
After getting a response, the app will update the UI with the latest data and cache them on the device.
Cache handled in other integrate.

### LINKED ISSUE #64: [Integrate] As a logged-in user, I can cache the survey list on the Home screen
## Why

- We need to cache the survey list to have a better UX.
 
## AC

- The surveys are cached on the device upon successful API retrieval, and on app relaunch, they are displayed from the cache; however, if there are differences with the cached data, they are displayed from the API.

### PR DESCRIPTION
**Pull request by nkhanh44**
- Close #31
- Close #64

## What happened 👀

- Added Hive for key-value database in dart and implemented survey_storage.
- Added HomeState to manage state of Home screen.
- Added HomeViewModel
- Implemented get_cached_surveys_use_case
- Added Unit tests

## Insight 📝

I faced the error when getting token from SecureStorage:

Screenshot 2023-08-23 at 12 32 13

So I just updated the casting at  `token_data_source` and `secure_storage_impl` and it works well now.

- One mistake that took me much time to find out is meta_response should use `decodedData` instead 🙈 .
- Load more and refreshing will be handled in other integrate.

## Proof Of Work 📹

- Success:

https://github.com/nimblehq/flutter-ic-khanh-thieu/assets/25881847/d570a8e9-9c69-4185-9ac0-63399bd4fd2b

https://github.com/nimblehq/flutter-ic-khanh-thieu/assets/25881847/9a45b8db-5ca9-4e4b-b4a4-bd8d502fba85

- Failure:

https://github.com/nimblehq/flutter-ic-khanh-thieu/assets/25881847/c36b43d0-496d-471c-be33-308af9c3ca20

### CODE REVIEW

File: lib/screens/home/home_view_model.dart [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+  void loadCachedSurveys() async {
+    final result = await _getCachedSurveysUseCase.call();
+    if (result is Success<List<SurveyModel>>) {
+      final cachedSurveys =
+          result.value.map((survey) => survey.toSurveyModel()).toList();
+      _surveys.add(cachedSurveys);
+      state = const HomeState.loadCachedSurveysSuccess();
+    }
+  }
```
Comment (Thieurom) [inline lib/screens/home/home_view_model.dart:65]: IMO, we shouldn't expose the `loadCachedSurveys` to the outside. When the view calls `loadSurveys()` on the view model, it'll return (a stream) list of survey items. The view doesn't have to know whether the list is cached or from remote. It just uses that list to build the UI.
```Dart
void loadSurveys() async {
  await _loadSurveysFromCache();
  await _loadSurveysFromRemote();
}
```

**Review by Thieurom — COMMENTED** [review]
This PR is quite large. These are some initial comments 🙏

File: lib/screens/home/home_view_model.dart (hunk 1 above)
Comment (nkhanh44) [reply lib/screens/home/home_view_model.dart]: Thanks, that makes sense 👍  I updated all at https://github.com/nimblehq/flutter-ic-khanh-thieu/pull/65/commits/8003a9340e50fa6d46bdc47719418e28e9359efb

**Review by doannimble — COMMENTED** [review]
Good work, rest lgtm 👍

File: lib/screens/home/home_pages_widget.dart [hunk 2]
```diff
@@ -1,49 +1,74 @@
...
+                  fit: BoxFit.cover,
+                  width: double.infinity,
+                  height: double.infinity,
                 ),
               ),
+              Column(
+                children: [
+                  const Spacer(),
+                  SafeArea(
+                    bottom: true,
```
[the remark this answers is not in the document]
Comment (nkhanh44) [reply lib/screens/home/home_pages_widget.dart]: Thanks 🙏 I updated all at https://github.com/nimblehq/flutter-ic-khanh-thieu/pull/65/commits/8b31157e9e11bf9fcbe2ff0b33ed511deb11cda8

**Review by doannimble — APPROVED** [review]
lgtm now 👍

**Review by Thieurom — APPROVED** [review]

**Pull request merged**